File: src/core/validators/InterchangeExtension/InterchangeExtensionMustNotDuplicateInterchangeElementName.cs
using System.Linq;
using MetaEd.Grammar.Antlr;

namespace MetaEd.Core.Validator.InterchangeExtension
{
    public class InterchangeExtensionMustNotDuplicateInterchangeElementName : ValidationRuleBase<MetaEdGrammar.InterchangeExtensionContext>
    {
        private readonly ISymbolTable _symbolTable;

        public InterchangeExtensionMustNotDuplicateInterchangeElementName(ISymbolTable symbolTable)
        {
            _symbolTable = symbolTable;
        }

        private static string[] DuplicateInterchangeElements(MetaEdGrammar.InterchangeExtensionContext context)
        {
            var interchangeElements = context.interchangeExtensionComponent().interchangeElement().Select(x => x.ID().GetText());

            //group and filter duplicates
            return interchangeElements.GroupBy(x => x).Where(group => group.Count() > 1).Select(group => group.Key).ToArray();
        }

        public override bool IsValid(MetaEdGrammar.InterchangeExtensionContext context)
        {
            return !DuplicateInterchangeElements(context).Any();
        }

        public override string GetFailureMessage(MetaEdGrammar.InterchangeExtensionContext context)
        {
            var identifier = context.extendeeName().GetText();
            var duplicateInterchangeElements = DuplicateInterchangeElements(context);

            return string.Format("Interchange additions '{0}' declares duplicate interchange element{2} '{1}'.", identifier, string.Join("', '", duplicateInterchangeElements), duplicateInterchangeElements.Count() > 1 ? "s" : string.Empty);
        }
    }
}

File: src/core/validators/InterchangeExtension/InterchangeExtensionMustNotDuplicateInterchangeElementName.ts
import { ValidationRuleBase } from "../ValidationRuleBase";
import {ISymbolTable} from '../SymbolTable'
export class InterchangeExtensionMustNotDuplicateInterchangeElementName extends ValidationRuleBase<MetaEdGrammar.InterchangeExtensionContext>
{
    private symbolTable: ISymbolTable;
    constructor(symbolTable: ISymbolTable) {
        super();
        this.symbolTable = symbolTable;
    }
    private static duplicateInterchangeElements(context: MetaEdGrammar.InterchangeExtensionContext): string[] {
        let interchangeElements = context.interchangeExtensionComponent().interchangeElement().map(x => x.ID().getText());
        return interchangeElements.GroupBy(x => x).filter(group => group.Count() > 1).map(group => group.Key).ToArray();
    }
    public isValid(context: MetaEdGrammar.InterchangeExtensionContext): boolean {
        return InterchangeExtensionMustNotDuplicateInterchangeElementName.duplicateInterchangeElements(context).length == 0;
    }
    public getFailureMessage(context: MetaEdGrammar.InterchangeExtensionContext): string {
        let identifier = context.extendeeName().getText();
        let duplicateInterchangeElements = InterchangeExtensionMustNotDuplicateInterchangeElementName.duplicateInterchangeElements(context);
        return `Interchange additions '${identifier}' declares duplicate interchange element{duplicateInterchangeElements.length > 1 ? "s" : ""} '${duplicateInterchangeElements.join(', ')}'`;
    }
}
